xfer/app_client.go

```diff
@@ -0,0 +1,154 @@
+}
+
+type appClient struct {
+	ProbeConfig
+
+	quit     chan struct{}
+	target   string
+	insecure bool
+	client   http.Client
+}
+
+func NewAppClient(pc ProbeConfig, hostname, target string) (AppClient, error) {
+	httpTransport, err := getHTTPTransport(hostname, pc.Insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	return &appClient{
+		ProbeConfig: pc,
+		quit:        make(chan struct{}),
+		target:      target,
+		client: http.Client{
+			Transport: httpTransport,
+		},
+	}, nil
+}
+
+func (c *appClient) Stop() {
+	close(c.quit)
+}
+
+// AuthorizationHeader returns a value suitable for an HTTP Authorization
+// header, based on the passed token string.
+func AuthorizationHeader(token string) string {
+	return fmt.Sprintf("Scope-Probe token=%s", token)
+}
+
+func (c *appClient) authoriseHeaders(headers http.Header) {
+	headers.Set("Authorization", AuthorizationHeader(c.Token))
+	headers.Set(ScopeProbeIDHeader, c.ProbeID)
+}
+
+func (c *appClient) authorizedRequest(method string, urlStr string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest(method, urlStr, body)
+	if err == nil {
+		c.authoriseHeaders(req.Header)
+	}
+	return req, err
+}
+
+func (c *appClient) Details() (Details, error) {
+	result := Details{}
+	req, err := c.authorizedRequest("GET", sanitize.URL("", 0, "/api")(c.target), nil)
+	if err != nil {
+		return result, err
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return result, err
+	}
+	defer resp.Body.Close()
+	return result, json.NewDecoder(resp.Body).Decode(&result)
+}
+
+func (c *appClient) Capabilities(handler CapabilityHandler) error {
+	dialer := websocket.Dialer{}
+	headers := http.Header{}
+	c.authoriseHeaders(headers)
+	// TODO(twilkie) need to update sanitize to work with wss
+	url := sanitize.URL("ws://", 0, "/api/capabilities")(c.target)
+	conn, _, err := dialer.Dial(url, headers)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	incoming, outgoing := make(chan Request), make(chan Response)
+	quit := make(chan struct{})
+
+	go func(c *websocket.Conn) {
+		defer close(quit)
+
+		for { // just discard everything the browser sends
+			_, reader, err := c.NextReader()
+			if err != nil {
+				log.Printf("Error reading message from %s: %v", err)
+				return
+			}
+
+			var request Request
+			if err := json.NewDecoder(reader).Decode(&request); err != nil {
+				log.Printf("Error parsing message from %s: %v", err)
+				return
+			}
+
+			incoming <- request
+		}
+	}(conn)
+
+	go func(c *websocket.Conn) {
+		defer close(quit)
+
+		for response := range outgoing {
+			//if err := conn.SetWriteDeadline(time.Now().Add(websocketTimeout)); err != nil {
+			//	log.Printf("Error SetWriteDeadline for %s: %v", err)
+			//	return
+			//}
+			if err := conn.WriteJSON(response); err != nil {
+				log.Printf("Error sending response to %s: %v", err)
+				return
+			}
+		}
+	}(conn)
+
+	for {
+		select {
+		case req := <-incoming:
+			// TODO process more than one request at once.
```

Comment:
Yeah needs more thoughts.  The request / responses can be interleaved on the wire, so thats not a problem.  Otherwise yeah, handling multiple requests at once might be a challenge.